File: BFS/BFS_intro_2.cpp
#include <bits/stdc++.h>
using namespace std;

void bfs(vector<vector<int>> &edges, int n)
{
    vector<unordered_set<int>> graph(n);
    for(auto &a:edges)
        graph[a[0]].insert(a[1]),
        graph[a[1]].insert(a[0]);

    queue<int> q(deque<int>{0});
    vector<bool> visited(n);
    vector<int> label(n);
    visited[0] = true;
    int cur_node;

    for(int qsize = 1; qsize != 0; qsize = q.size())
    {
        for(int i=0; i<qsize; i++)
        {
            cur_node = q.front();
            q.pop();
            cout << cur_node <<  " ";    // current node 
            for(auto &v: graph[cur_node])
            {
                if(!visited[v])
                {
                    q.push(v);
                    visited[v] = true;
                    label[v] = label[cur_node] + 1;
                }
            }   
        }
    }
    cout << endl << endl << "Node   " << "Label No" << endl;
    for(int i=0; i<n; i++)
        cout << " " << i << "   :   " << label[i] << endl; 
}

int main()
{
    vector<vector<int>> edges = {{0,1},{0,2},{0,9},{1,5},{1,6},{2,4},{4,7},{6,8}, {7,3}};
    bfs(edges, 10);
    return 0;
}

File: BFS/BFS_intro_2.py
from queue import Queue

def bfs(edges:list[list[int]], n:int) -> None:
    graph = [ set() for _ in range(n) ]
    for (u,v) in edges:
        graph[u].add(v)
        graph[v].add(u)

    q = Queue()
    visited = [False] * n
    level = [0] * n
    q.put(0)
    visited[0] = True

    while not q.empty():
        for _ in range(q.qsize()):
            cur_node = q.get()
            print(cur_node, end=" ")
            for v in graph[cur_node]:
                if not visited[v]:
                    q.put(v)
                    visited[v] = True
                    level[v] = level[cur_node] + 1
                
    print()
    print('\n'.join( f'{i}  :  {v}' for i,v in enumerate(level)))

def main():
    edges = [[0,1],[0,2],[0,9],[1,5],[1,6],[2,4],[4,7],[6,8],[7,3]]
    bfs(edges, 10)

if __name__ == "__main__":
    main()